cmd/erasure-healfile.go

```diff
@@ -36,48 +50,86 @@ func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string,
 			f.Checksums[i] = checksums[i]
 		}
 	}
-	blocks := make([][]byte, len(s.disks))
+
+	// Scan part files in disk, block-by-block reconstruct it and
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for ; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
+		var buffer []byte
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			if disk == OfflineDisk {
+				continue
+			}
+			// allocate buffer only when needed - when
+			// reads fail, the buffer can be reused
+			if int64(len(buffer)) != chunksize {
+				buffer = make([]byte, chunksize)
+			}
+			if !verifiers[i].IsVerified() {
+				_, err = disk.ReadFileWithVerify(volume, path,
+					chunkOffset, buffer, verifiers[i])
+			} else {
+				_, err = disk.ReadFile(volume, path,
+					chunkOffset, buffer)
+			}
+			if err != nil {
+				// LOG FIXME: add a conditional log
+				// for read failures, once per-disk
+				// per-function-invocation.
+				continue
+			}
+
+			// read was successful, so set the buffer as
+			// blocks[i], and reset buffer to nil to force
+			// allocation on next iteration
+			blocks[i], buffer = buffer, nil
```

Comment:
This does not work. The problem is that this solution will allocate new buffers all the time. `buffer` is always set to nil so `int64(len(buffer)) != chunksize` always fails and the buffer gets allocated again.

Reply:
It does work. See https://play.golang.org/p/abVips4qTA

Also, buffer is set to nil only when the read succeeds. When it fails, the buffer is reused. As noted in the comments.

Reply:
I'm not saying it does not work in the sense of correctness - I'm saying it does not reduce the number of allocations.
The reason is that your solution allocates always a new block on a successful read - which should happen much more often than a fail (the disks are already filtered: online | offline). See this simpler comparison of both solutions: [https://play.golang.org/p/0t5FxQ0k5n](https://play.golang.org/p/0t5FxQ0k5n)

Reply:
@aead Thank you for the detailed explanation. However, your solution appears to have a bug. When `blocks` is passed to `s.ErasureDecodeDataAndParityBlocks()`, it is required to have the following structure - any `s.dataBlocks` number of slices in `blocks` must be containing valid data (therefore not set to nil), all the other slices are required to be set to nil to indicate that they are missing blocks (i.e. erasure reconstruction will allocate and reconstruct them) - [(doc)](https://github.com/donatello/minio/blob/429aa1e0ab3b736268ae12365ba6fda3a8428d67/vendor/github.com/klauspost/reedsolomon/reedsolomon.go#L50).

Now, read failures can happen to any subset of the online disks for any blocks (specially, even between blocks in the same object - for e.g. because disk errors can happen randomly even to online disks). 

Consider a situation where after reading chunk number `i` successfully, the file-reading loop breaks because `s.dataBlocks` number of chunks have been collected. At this point, any of the slices in `blocks` with slice index > `i` may not be nil (due to sharing `blocks` across iterations of the outer for-loop`). If this happens, erasure reconstruction will receive garbage and return garbage.

Due to this possibility, in my solution, `blocks` is not shared across iterations of the outer for-loop. A difficulty in further reducing memory allocs is that the lower-level erasure reconstruction function always allocates the blocks it is reconstructing. Given this limitation the only allocations that can be saved are the ones due to file-read failures.

Reply:
@donatello Yes, I see the problem with Reconstruct. In general correct - we need to mark all shards as missing. I looked at the reedsolomon code and there is actually a [problem](https://github.com/klauspost/reedsolomon/blob/master/reedsolomon.go#L685) - it should check whether the capacity is large enough to hold the shard. I'll open an issue and send a patch.
Without this library fix your solution is completely fine and solves the issues. As soon we get this fix into reedsolomon we can do the following we can optimize HealFile correctly - for now there is nothing we can do about.
Thanks for the hint :+1: